sound/soc/codecs/rt711-sdca-sdw.c

```diff
@@ -378,6 +378,8 @@ static int __maybe_unused rt711_sdca_dev_suspend(struct device *dev)
 
 	regcache_cache_only(rt711->regmap, true);
 	regcache_cache_only(rt711->mbq_regmap, true);
+	regcache_mark_dirty(rt711->regmap);
+	regcache_mark_dirty(rt711->mbq_regmap);
```

Comment:
that's not the right fix, we need to extend the existing one with this

````
	if (rt711->first_hw_init) {
		regcache_cache_bypass(rt711->regmap, false);
		regcache_mark_dirty(rt711->regmap);
>>		regcache_cache_bypass(rt711->regmap_mbq, false);
>>		regcache_mark_dirty(rt711->regmap_mbq);
	} else
		rt711->first_hw_init = true;
````

Reply:
> that's not the right fix, we need to extend the existing one with this
> 
> ```
> 	if (rt711->first_hw_init) {
> 		regcache_cache_bypass(rt711->regmap, false);
> 		regcache_mark_dirty(rt711->regmap);
> >>		regcache_cache_bypass(rt711->regmap_mbq, false);
> >>		regcache_mark_dirty(rt711->regmap_mbq);
> 	} else
> 		rt711->first_hw_init = true;
> ```

I did it in the next patch. I think we need both because there are two scenarios. 
1. Codec is unattached. In this case we need to do `rt711_sdca_io_init`
2. Common suspend/resume case. In this case we don't need to do `rt711_sdca_io_init`, but need to sync some registers since they will be restore to default when we do bus reset.

Reply:
@bardliao I think we are not aligned on the io_init case. This function will be used when we reattach, and there's no difference between the codec resetting on its own and the bus reset. The codec becoming un-attached can be for different reasons with different side effects, but there is nothing that tells us what the source of the reset was. It's a miss in the SoundWire spec and as a result software has to assume that all the state was lost every time the device re-attaches.

Reply:
@plbossart So the problem is actually `sdw_update_slave_status(SDW_SLAVE_UNATTACHED)` is never triggered in the reattach case. See [dmesg.txt](https://github.com/thesofproject/linux/files/6491343/dmesg.txt) with [bard.diff.txt](https://github.com/thesofproject/linux/files/6491348/bard.diff.txt). Therefore, [rt711->hw_init = false;](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/codecs/rt711-sdca-sdw.c#L139)  and [rt711_sdca_io_init](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/codecs/rt711-sdca-sdw.c#L164) will not be invoked
I will look into it.

Reply:
@plbossart I got why `sdw_update_slave_status(SDW_SLAVE_UNATTACHED)` is never triggered now. We call `sdw_modify_slave_status`  without `sdw_update_slave_status` in `sdw_clear_slave_status`. As a result, `sdw_update_slave_status(SDW_SLAVE_UNATTACHED)` will never be invoked elsewhere due to the
status is not changed. Please see https://github.com/thesofproject/linux/pull/2928.
However, with #2928 and this PR updated, I see 
```
[   76.370376] rt711-sdca sdw:0:025d:0711:01: DPN_PortCtrl register write failed for port 3
[   76.370380] soundwire sdw-master-0: Program transport params failed: -61
[   76.370382] soundwire sdw-master-0: Program params failed: -61
```
[dmesg.txt](https://github.com/thesofproject/linux/files/6493665/dmesg.txt)
I think it could be something wrong with  `rt711_sdca_io_init` @shumingfan FYI